ユーザー管理機能

nemoimoi opened the pull request:
# What
ユーザー管理機能の実装

# Why
新規登録やログイン、ログアウトなどのユーザー管理機能を設定するため

■必要な情報を適切に入力して「会員登録」ボタンを押すと、ユーザーの新規登録ができる動画
https://gyazo.com/a6e2134cc0d02089d8fd1909ec2e385c

■入力に問題のある状態で「会員登録」ボタンが押された場合、ユーザーの新規登録はできず、エラーメッセージが出力される動画
https://gyazo.com/2ea1d2c9dd2b116bc680ef24164dbbcb

■必要な情報を適切に入力して「ログイン」ボタンを押すと、ユーザーのログインができる動画
https://gyazo.com/008c7df50445dfc91f2cb535c719aee4

■入力に問題のある状態で「ログイン」ボタンが押された場合、ユーザーのログインはできず、エラーメッセージが出力される動画
https://gyazo.com/e815525896d15d9f1615342e24e37cfa

■ログアウトに成功する動画
https://gyazo.com/9beaf727e2ec468a46daa256850919d3

■テストに成功する画像
https://gyazo.com/62e5912122bd69e2b21d030157e01815

nemoimoi committed:
ユーザー管理機能

・新規登録の実装
・ログイン実装
・ログアウト実装
・テストコード完了

shodai-ito commented on app/models/user.rb:
```diff
@@ -0,0 +1,20 @@
+    validates :first_name, presence:true
+  end
+
+  with_options presence: true, format: { with: /\A[ァ-ヶ]+\z/} do
```
氏名（カナ）の正規表現を`/\A[ァ-ヶー]+\z/`に変更しましょう！
現在の表記ですと長音（ー）が含まれていないため、メアリーなど英語名の方が登録できない状態です。

shodai-ito commented on app/views/devise/sessions/new.html.erb:
```diff
@@ -0,0 +1,36 @@
+<%= render "shared/second-header"%>
+
+<%= form_with class: 'registration-main', local: true do |f| %>
+<%= form_with model:resource, as: resource_name, url: session_path(resource_name), html:{data: { turbo:false } } do |f| %>
```
form_withの記述が重複しているため、コードの整理をお願いいたします！

shodai-ito commented on app/views/devise/registrations/new.html.erb:
```diff
@@ -0,0 +1,108 @@
+
+<%# 「モデル名」にはUserモデルであれば@userを渡しましょう。「新規登録機能へのパス」は、devise導入後にrails routesを実行してdevise/registrations#createへのパスを確認し、記載してください。 %>
+<%= form_with model: @user, url: user_registration_path, class: 'registration-main', local: true do |f| %>
+<%# //「モデル名」にはUserモデルであれば@userを渡しましょう。「新規登録機能へのパス」は、devise導入後にrails routesを実行してdevise/registrations#createへのパスを確認し、記載してください。 %>
```
実装`済みの`案内用文は削除しましょう！

コードレビューは一旦完成したコードに対して行うものであるのと、企業様に提出する際に実装案内のコメントが残っていることは好ましくないためです。
もし学習メモとして必要であれば、第三者でもわかるように「残す目的」「なぜ必要か」分かるよう、`「学習メモ」`等分かりやすいコメントとして書いておくことをおすすめいたします。

■補足
他にも不要な案内文があれば、削除しましょう。

shodai-ito commented on app/views/devise/registrations/new.html.erb:
```diff
@@ -0,0 +1,108 @@
+  </div>
+
+  <%# インスタンスを渡して、エラー発生時にメッセージが表示されるようにしましょう。%>
+  <%# render 'shared/error_messages', model: f.object %>
```
ユーザー登録に失敗した際、エラーメッセージが表示されていないかと存じます。
エラーメッセージが表示されるように実装をお願いします。

shodai-ito commented on spec/factories/users.rb:
```diff
@@ -0,0 +1,13 @@
+  factory :user do
+    nickname {Faker::Internet.username}
+    email {Faker::Internet.email}
+    password {Faker::Internet.password(min_length: 6)}
```
Factorybotのパスワードの値を修正しましょう！

Fakerだけの場合、稀に「数字のみ」や「英字のみ」のパスワードが生成されてしまい、テストが失敗することが考えられます。

要件書に「パスワードは、半角英数字混合での入力が必須であること」とありますので、パスワードが必ず英数混合になるように記述をしてみましょう。

◎方法1
英数字混合のpasswordを想定した上でFakerを使用する場合は、例えば

```ruby
    password              { '1a' + Faker::Internet.password(min_length: 6) }
```
のように、確実にpasswordに英数字が含まれるように指定することが可能です。

◎方法2
Fakerを用いずに{"aaaaa1"}と直接値を入れて実装することも可能です。

■直接値を入れているやり方
```ruby
コース：応用
タイトル：RSpec〜効率的にテストコードを書こう〜

FactoryBotを使用してインスタンス生成を共通化しよう
```

shodai-ito commented on spec/models/user_spec.rb:
```diff
@@ -0,0 +1,81 @@
+        @user.encrypted_password = 'n12345'
+        @user.valid?
+        expect(@user.errors.full_messages).to include("Password confirmation doesn't match Password")
+      end
```
パスワードの半角英数字混合を確認するために3つのテストで確認しましょう！

```ruby
①パスワードが半角英字のみでは登録できないこと
②パスワードが半角数字のみでは登録できないこと
③パスワードが全角では登録できないこと
```

shodai-ito commented on spec/models/user_spec.rb:
```diff
@@ -0,0 +1,81 @@
+        @user.email = 'testmail'
+        @user.valid?
+        expect(@user.errors.full_messages).to include('Email is invalid')
+      end
```
名前について正規表現を実装しておりますので、それらについてもテストを行いましょう！
```
 ・姓（全角）に半角文字が含まれていると登録できない
 ・名（全角）に半角文字が含まれていると登録できない
 ・姓（カナ）にカタカナ以外の文字（平仮名・漢字・英数字・記号）が含まれていると登録できない
 ・名（カナ）にカタカナ以外の文字（平仮名・漢字・英数字・記号）が含まれていると登録できない
```

nemoimoi replied:
レビューありがとうございます！
修正いたします！

shodai-ito reviewed (commented):
レビュー依頼ありがとうございます！
プルリクエストの内容を確認させていただきました。

修正点をコメントさせて頂きましたので、ご確認をお願い致します。

【次回プルリクエスト提出する際の注意点】
```ruby
①修正が終わりましたら、`コミットとプッシュ`をして、`プルリクエストに修正内容が反映されているか確認`しましょう。
②既に頂いている挙動確認の画像や動画が修正範囲だった場合、最新の画像や動画の提出をお願い致します。
③googleフォームの送信をお願いいたします。
```


ご不明点や疑問点ございましたら、チャット・コネクトメンターにご相談ください！
どうぞよろしくおねがいします。

nemoimoi committed:
ユーザー管理機能デバック

・カタカナの正規表現を変更
・form_withの記述を整理
・不要なコメントを削除
・エラーメッセージが表示されるように修正
・Factorybotのpasswordの値を半角英数混合になるように修正
・passwordが半角英数字混合になっているか確認のためテストコード追加
・名前の登録について正規表現を加味したテストコード追加

nemoimoi commented:
# What
ユーザー管理機能デバックを行いました。
# Why
修正が必要な個所と対応できていない箇所があったためレビューを元に以下内容を変更及び追記致しました。
・カタカナの正規表現を変更
・form_withの記述を整理
・不要なコメントを削除
・エラーメッセージが表示されるように修正
・Factorybotのpasswordの値を半角英数混合になるように修正
・passwordが半角英数字混合になっているか確認のためテストコード追加
・名前の登録について正規表現を加味したテストコード追加

■入力に問題のある状態で「会員登録」ボタンが押された場合、ユーザーの新規登録はできず、エラーメッセージが出力される動画
https://gyazo.com/be34d5ff7898ef9a7e7ec2b964950031

■入力に問題のある状態で「ログイン」ボタンが押された場合、ユーザーのログインはできず、エラーメッセージが出力される動画
https://gyazo.com/a500d825c1cecde0d09834b0847075e2

■テストに成功する画像
https://gyazo.com/4e898ec36597532729972e29ae565eef

nemoimoi commented:
レビュー依頼をフォームより行いました！
よろしくお願いいたします！

wad-sho commented on app/views/shared/_header.html.erb:
```diff
@@ -16,12 +16,13 @@
+        <li><%= link_to 'ログイン', new_user_session_path, class: "login" %></li>
+        <li><%= link_to '新規登録', new_user_registration_path, class: "sign-up" %></li>
+      <% end %>
       <%# //deviseを導入できたら、ログインの有無で表示が変わるように分岐しましょう%>
```
「〜しましょう」で終わる実装案内コメントは実装完了後に削除しましょう。
企業提出時に実装案内コメントが残っていないようにするためです。

nemoimoi replied:
レビューありがとうございます！
コメント削除漏れ失礼いたしました。
削除いたしました。

wad-sho reviewed (commented):
実装お疲れ様です！レビューのご依頼ありがとうございます。

細かい点ですがコメントいたしましたので、ご確認・修正の後、再度レビューのご依頼お願いいたします！

nemoimoi committed:
ユーザー管理機能デバック２

・不要コメント削除

nemoimoi commented:
# What
不要コメントの削除を行いました。
# Why
削除予定の不要コメントが残っていたため削除致しました。

nemoimoi commented:
不要コメントが残っていたようで失礼しました。。
フォームよりレビュー依頼を行いましたのでよろしくお願いします！

shodai-ito commented:
レビュー依頼ありがとうございます！

こちら`LGTM`とさせていただきます！実装お疲れさまでした。

【こちらは次回の機能に入る際の注意点です】
```ruby
プルリクエストのMerge、ローカルのmainブランチプルを忘れずに行い、ブランチを切ってから次の実装に取り組んで頂ければと思います。
```

ご不明点や疑問点ございましたら、チャット・コネクトメンターにご相談ください！
次は「出品機能」かと思われます！引き続き頑張りましょう！応援しております！
